src/coreclr/tools/Common/Compiler/ObjectWriter/WasmNative.cs

```diff
@@ -0,0 +1,175 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+
+namespace ILCompiler.ObjectWriter
+{
+    public enum WasmSectionType
+    {
+        Custom = 0,
+        Type = 1,
+        Import = 2,
+        Function = 3,
+        Table = 4,
+        Memory = 5,
+        Global = 6,
+        Export = 7,
+        Start = 8,
+        Element = 9,
+        Code = 10,
+        Data = 11,
+        DataCount = 12,
+        Tag = 13,
+    }
+
+    public static class DummyValues
+    {
+        // Wasm function signature for (func (params i32) (result i32))
+        public static WasmFuncType CreateWasmFunc_i32_i32()
+        {
+            return new WasmFuncType(
+                paramTypes: new([WasmValueType.I32]),
+                returnTypes: new([WasmValueType.I32])
+            );
+        }
+    }
+
+    public enum WasmValueType
+    {
+        I32 = 0x7F,
+        I64 = 0x7E,
+        F32 = 0x7D,
+        F64 = 0x7C
+    }
+
+    public static class WasmValueTypeExtensions
+    {
+        public static string ToString(this WasmValueType valueType)
+        {
+            return valueType switch
+            {
+                WasmValueType.I32 => "i32",
+                WasmValueType.I64 => "i64",
+                WasmValueType.F32 => "f32",
+                WasmValueType.F64 => "f64",
+                _ => "unknown",
+            };
+        }
+    }
+
+    public struct WasmResultType : IEquatable<WasmResultType>
+    {
+        private readonly WasmValueType[] _types;
+        public readonly Span<WasmValueType> Types => _types;
+
+        public WasmResultType(WasmValueType[] types)
+        {
+            _types = types;
+        }
+
+        public bool Equals(WasmResultType other) => Enumerable.SequenceEqual(_types, other._types);
+        public override bool Equals(object obj)
+        {
+            return obj is WasmResultType other && Equals(other);
+        }
+
+        public override int GetHashCode()
+        {
+            if (_types == null || _types.Length == 0)
+                return 0;
+
+            int code = _types[0].GetHashCode();
+            for (int i = 1; i < _types.Length; i++)
+            {
+                code = HashCode.Combine(code, _types[i].GetHashCode());
+            }
+
+            return code;
+        }
+
+        public int EncodeSize()
+        {
+            uint sizeLength = DwarfHelper.SizeOfULEB128((ulong)_types.Length);
+            return (int)(sizeLength + (uint)_types.Length);
+        }
+
+        public int Encode(Span<byte> buffer)
+        {
+            DwarfHelper.WriteULEB128(buffer, (ulong)_types.Length);
+            uint sizeLength = DwarfHelper.SizeOfULEB128((ulong)_types.Length);
+            for (int i = 0; i < _types.Length; i++)
+            {
+                buffer[(int)sizeLength + i] = (byte)_types[i];
```

Comment:
AFAIK types are actually encoded as signed LEB128 and in some cases (block types) can be the index of a function type signature to represent multiple return values. It's probably fine to write bytes for now, but maybe we should leave a comment somewhere pointing out that value types aren't really uint8's

Reply:
Also, WasmValueType is not currently constrained to having a basetype of byte, so an out of range value could be in `_types` and silently get truncated to uint8 here. I don't think you have to guard against that explicitly, but maybe give the enum a basetype?